pkg/agent/logrouter/logrouter.go

```diff
@@ -0,0 +1,203 @@
+
+package logrouter
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	agenttypes "github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/log/locallogger"
+	"github.com/kolide/launcher/pkg/log/ringlogger"
+	"github.com/kolide/launcher/pkg/log/teelogger"
+	"github.com/kolide/launcher/pkg/persistentring"
+)
+
+type teeloggerInt interface {
+	Add(loggers ...log.Logger)
+	Log(keyvals ...interface{}) error
+}
+
+type storeInt interface {
+	agenttypes.GetterSetterDeleterIterator
+}
+
+type ringLoggerInt interface {
+	Log(keyvals ...interface{}) error
+	GetAll() ([]map[string]any, error)
+}
+
+type logRouter struct {
+	loggerTee   teeloggerInt
+	systemTee   teeloggerInt
+	logRing     ringLoggerInt // Holds the super set of logs, in memory
+	persistRing ringLoggerInt // Only info and higher, persisted to a store
+}
+
+func New(systemLogger log.Logger) (*logRouter, error) {
+	// We always have exactly one in memory ring logger. This receives _all_ the logs. It's used to replay into new
+	// desinatations, and is expected to be a query target for recent logs
+	logRing, err := ringlogger.New(persistentring.NewInMemory(1000))
+	if err != nil {
+		return nil, fmt.Errorf("creating memory ringlogger: %w", err)
+	}
+
+	// Setup the internal logger, this rolls up all the internal destinations
+	loggerTee := teelogger.New(logRing)
+
+	// the systemLogger should copy anything it generates into the regular logger.
+	systemTee := teelogger.New(systemLogger, loggerTee)
+
+	return &logRouter{
+		loggerTee: loggerTee,
+		systemTee: systemTee,
+		logRing:   logRing,
+	}, nil
+}
+
+func (lr *logRouter) AddLogger(logger log.Logger) {
+	lr.loggerTee.Add(logger)
+}
+
+func (lr *logRouter) ReplaceSystemLogger(newSystemLogger log.Logger) {
+	newSystemTee := teelogger.New(newSystemLogger, lr.loggerTee)
+	lr.systemTee = newSystemTee
+}
+
+func (lr *logRouter) Log(keyvals ...interface{}) error {
+	return lr.loggerTee.Log(keyvals...)
+}
+
+func (lr *logRouter) Logger() log.Logger {
+	return lr.loggerTee
+}
+
+func (lr *logRouter) SystemLogger() log.Logger {
+	return lr.systemTee
+}
+
+func (lr *logRouter) GetRecentLogs() ([]map[string]any, error) {
+	if lr.persistRing == nil {
+		return nil, nil
+	}
+
+	return lr.persistRing.GetAll()
+}
+
+func (lr *logRouter) GetRecentDebugLogs() ([]map[string]any, error) {
+	return lr.logRing.GetAll()
+}
+
+// Replay will replay the logs from the memory ring to a new logger. Because we're serliazing everything down to a
+// string, handling the leveling is cumbersome.
+//
+// TODO: This should probably move to the underlying ringlogger, but we should also add an iterator for that.
+func (lr *logRouter) Replay(logger log.Logger) error {
+	logs, err := lr.logRing.GetAll()
+	if err != nil {
+		return fmt.Errorf("gettings logs to replay: %w", err)
+	}
+	for _, l := range logs {
+		var levelFn func(logger log.Logger) log.Logger
+
+		pairs := make([]any, 2+len(l))
```

Comment:
Preallocating the pairs array here and then appending leaves the the first few elements in the array nil and appends to the end of it.

So we should either keep track of the index and insert into the array or just make an empty array and append to it.

Here what the debugger shows when debugging the tests:

![image](https://github.com/kolide/launcher/assets/8415811/90764124-b928-4dc8-adfc-c670454c2cbb)

Reply:
Thanks for calling that out. I always _want_ to hint at length, to avoid extra allocations, but it probably doesn't work. I've removed it,